### PR TITLE
Reuse GenAPI GenerateReferenceAssemblySource target

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/corefx/issues/33981
Fixes https://github.com/dotnet/corefx/issues/31193
Requires https://github.com/dotnet/arcade/pull/1701

I decided to still recommend the target as the property together with a build/rebuild target is to noisy as it also updates the ref sources of referenced projects.

Commit 2 is just a manual darc update.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
@ericstj @wtgodbe PTAL

**Comment by ViktorHofer** [comment]
@dotnet-bot test Windows x64 Debug Build
@dotnet-bot test Windows x64 Release Build

**Comment by ViktorHofer** [comment]
@dotnet-bot test Windows x86 Release Build

**Comment by ViktorHofer** [comment]
The errors are all unrelated. OSX --> Jenkins exception, Windows --> EventLog https://github.com/dotnet/corefx/pull/33662#issuecomment-452371494

**Comment by ViktorHofer** [comment]
@dotnet-bot test OSX x64 Debug Build

File: src/Directory.Build.targets [hunk 1]
```diff
@@ -30,6 +24,24 @@
     <RestoreProjectStyle>None</RestoreProjectStyle>
   </PropertyGroup>
 
+  <!-- Settings for GenerateReferenceAssemblySource. -->
+  <PropertyGroup>
+    <_ExcludeAPIList>$(RepositoryEngineeringDir)DefaultGenApiDocIds.txt</_ExcludeAPIList>
+    <_LicenseHeaderTxtPath>$(RepositoryEngineeringDir)LicenseHeader.txt</_LicenseHeaderTxtPath>
+    <GenAPITargetPath>$([MSBuild]::NormalizePath('$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildProjectDirectory), Directory.Build.props))', 'ref', '$(TargetName).cs'))</GenAPITargetPath>
+    <GenAPILibPath>$(RefPath)</GenAPILibPath>
+    <GenAPIAdditionalParameters>--exclude-attributes-list "$(_ExcludeAPIList)" --header-file "$(_LicenseHeaderTxtPath)"</GenAPIAdditionalParameters>
```
Comment (ericstj) [inline src/Directory.Build.targets]: Consider reusing ApiCompatExcludeAttributeList, also consider making these actual extension points.

File: src/Directory.Build.targets [hunk 2]
```diff
@@ -30,6 +24,24 @@
...
+    <SkipGenAPITargetPathFileWrite>true</SkipGenAPITargetPathFileWrite>
+  </PropertyGroup>
+
+  <Target Name="AddSkipGetTargetFrameworkToProjectReferences" Condition="'@(ProjectReference)' != ''">
```
Comment (ericstj) [inline src/Directory.Build.targets]: Why did you move this target?

File: src/Directory.Build.targets [hunk 3]
```diff
@@ -1,5 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
+<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/Directory.Build.targets]: Yes sorry, I will undo these changes.

File: src/Directory.Build.targets [hunk 4]
```diff
@@ -30,6 +24,24 @@
     <RestoreProjectStyle>None</RestoreProjectStyle>
   </PropertyGroup>
 
+  <!-- Settings for GenerateReferenceAssemblySource. -->
+  <PropertyGroup>
+    <_ExcludeAPIList>$(RepositoryEngineeringDir)DefaultGenApiDocIds.txt</_ExcludeAPIList>
+    <_LicenseHeaderTxtPath>$(RepositoryEngineeringDir)LicenseHeader.txt</_LicenseHeaderTxtPath>
+    <GenAPITargetPath>$([MSBuild]::NormalizePath('$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildProjectDirectory), Directory.Build.props))', 'ref', '$(TargetName).cs'))</GenAPITargetPath>
+    <GenAPILibPath>$(RefPath)</GenAPILibPath>
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/Directory.Build.targets]: I think the reason was that I wanted to make it work by just setting the property `GenerateReferenceAssemblySource` (https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.GenAPI/build/Microsoft.DotNet.GenAPI.targets#L20) during the build. That crashed as one of the reference projects was a depproj file which does not have any references. By setting it to RefPath I was able to workaround this but the downside of using the property is that it also updates the ref cs files of the referenced projects.

File: src/Directory.Build.targets [hunk 5]
```diff
@@ -30,6 +31,15 @@
     <RestoreProjectStyle>None</RestoreProjectStyle>
   </PropertyGroup>
 
+  <!-- Settings for GenerateReferenceAssemblySource. -->
+  <PropertyGroup>
+    <_LicenseHeaderTxtPath>$(RepositoryEngineeringDir)LicenseHeader.txt</_LicenseHeaderTxtPath>
+    <GenAPITargetPath>$([MSBuild]::NormalizePath('$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildProjectDirectory), Directory.Build.props))', 'ref', '$(TargetName).cs'))</GenAPITargetPath>
```
Comment (ericstj) [inline src/Directory.Build.targets]: Consider reusing the convention defined in https://github.com/dotnet/corefx/blob/master/eng/resolveContract.targets.  Not sure if it's worth it but you may want to move all these properties to a props fine under eng.

Comment (wtgodbe) [reply src/Directory.Build.targets]: I like the idea of having a props file for these, that way we can share it later if we wind up needing to set stuff for APICompat/GenFacades/etc

File: src/Directory.Build.targets (hunk 4 above)
[the remark this answers is not in the document]
Comment (ericstj) [reply src/Directory.Build.targets]: That's a different problem though.  We could avoid running the target for projects which don't have reference assembly projects.

We can also keep the workflow of manually running the target working by ensuring we define DependsOnTargets to run ResolveReferences.  That way we don't need to use the property to enable this.

**Comment by ericstj** [comment]
@ViktorHofer are you still working on this one?  /cc @ahsonkhan who's recently had some experience here.

**Comment by karelz** [comment]
I will close it for now as it seems abandon for a while. @ViktorHofer feel free to reopen it when you are ready to push it forward. Thanks!

**Comment by ViktorHofer** [comment]
@ericstj I currently don't have time to finish this. If you want, you can take it over.
